# Homework 4

> I pledge my Honor that I have abided by the Stevens Honor System. - Joshua Schmidt

## Question 1

Consider again the example application of Bayes rule in Section 6.2.1 of Tom Mitchell’s textbook (or slide page 6 of Lecture 6). Suppose the doctor decides to order a second laboratory test for the same patient, and suppose the second test returns a positive result as well. What are the posterior probabilities of cancer and -cancer respectively following these two tests? Assume that the two tests are independent.

If the two tests are independent, the posterior probabilities for cancer and -cancer are the same following the two tests as they are for the first test. Therefore with a propbility of a single person in the population having this cancer of 0.008, this results in: P(cancer) = 0.008, P(-cancer) = 0.992. P(+|cancer) = 0.98, P(-|cancer) = 0.02. P(+|-cancer) = 0.03, P(-|-cancer) = 0.97.

$P(-|-cancer) = \frac{0.98 \cdot 0.008}{0.98 \cdot 0.008 + 0.992 \cdot 0.03} = 0.97$

## Question 2

In lecture 6 (page 28-29) we used the Naïve Bayes Algorithm to predict a new instance based on a dataset with 14 examples (page 28). If we only have 12 examples as shown below, what is the prediction results for the same new instance? Show your calculation.

$P(y) \cdot P(sun | y) \cdot P(cool | y) \cdot P(high | y) \cdot P(strong | y)$  
$\frac{8}{12} \cdot \frac{2}{8} \cdot \frac{3}{8} \cdot \frac{3}{8} \cdot \frac{3}{8}$  
$=0.0087890625$

$P(n) \cdot P(sun | n) \cdot P(cool | n) \cdot P(high | n) \cdot P(strong | n)$  
$\frac{4}{12} \cdot \frac{3}{4} \cdot \frac{1}{4} \cdot \frac{3}{4} \cdot \frac{2}{4}$  
$=0.0234375$

## Question 3

Consider a 2 layer feedforward ANN with 2 inputs a and b, 1 hidden unit c, and 1 output unit d. Run 2 iterations and show the output.

activation function: $\sigma(y)=\frac{1}{1+e^{-y}}$

training iteration 1:  
$output_c = \sigma(0.1 \cdot 1 + 0.1 \cdot 0 + 0.1 \cdot 1) = \sigma(0.2) = 0.55$  
$output_d = \sigma(0.1 \cdot 0.55 + 0.1 \cdot 1) = \sigma(0.155) = 0.54$  

error for 2 neurons:  
$error_d = 0.54 \cdot (1 - 0.54) \cdot (1 - 0.54) = 0.115$  
$error_c = 0.55 \cdot (1 - 0.55) \cdot 0.1 \cdot 0.115 = 0.0028$  

deltas:  
$\delta_{c0} = 0.3 \cdot 0.0028 \cdot 1 = 0.00085$  
$\delta_{ca} = 0.3 \cdot 0.0028 \cdot 1 = 0.00085$  
$\delta_{d0} = 0.3 \cdot 0.115 \cdot 1 = 0.034$  
$\delta_{dc} = 0.3 \cdot 0.115 \cdot 0.55 = 0.019$  
$\delta_{cb} = 0.3 \cdot 0.0028 \cdot 0 = 0$  

new weights:  
$w_{c0}=0.1 + 0.00085 = 0.10085$  
$w_{ca}=0.1 + 0.00085 = 0.10085$  
$w_{d0}=0.1 + 0.034 = 0.134$  
$w_{dc}=0.1 + 0.019 = 0.119$  
$w_{cb}=0.1 + 0 = 0.1$  

tranining iteration 2:  
$output_c = \sigma(0.10085 \cdot 1 + 0.10085 \cdot 0 + 0.1 \cdot 1) = \sigma(0.20085) = 0.55$  
$output_d = \sigma(0.119 \cdot 0.55 + 0.134 \cdot 1) = \sigma(0.1996) = 0.5497$  

error for 2 neurons:  
$error_d = 0.5497 \cdot (1 - 0.5497) \cdot (1 - 0.5497) = -0.1361$  
$error_c = 0.55 \cdot (1 - 0.55) \cdot 0.119 \cdot (-0.1361) = -0.0004$  

deltas:  
$\delta_{c0} = 0.3 \cdot (-0.004) \cdot 1 + 0.9 \cdot 0.00085 = -0.0004$  
$\delta_{ca} = 0.3 \cdot (-0.004) \cdot 0 + 0.9 \cdot 0.00085 = -0.00086$  
$\delta_{d0} = 0.3 \cdot (-0.1361) \cdot 1 + 0.9 \cdot 0.034 = -0.01$  
$\delta_{dc} = 0.3 \cdot (-0.1361) \cdot 0.55 + 0.9 \cdot 0.019 = -0.0055$  
$\delta_{cb} = 0.3 \cdot (-0.004) \cdot 1 + 0.9 \cdot 0 = -0.0012$  

new weights:  
$w_{c0}=0.10085 - 0.0004 = 0.10085$  
$w_{ca}=0.10085 + 0.00086 = 0.1016$  
$w_{d0}=0.11342 - 0.01 = 0.1242$  
$w_{dc}=0.1189 - 0.0055 = 0.1134$  
$w_{cb}=0.1 - 0.0012 = 0.0988$  

## Question 4

imports

In [3]:
import pandas as pd
import numpy as np
from sklearn.feature_extraction import DictVectorizer
from sklearn.model_selection import train_test_split
from sklearn.neural_network import MLPClassifier

random_state = 42

In [6]:
titanic_df = pd.read_csv('./titanic.csv', index_col=0)
titanic_df.head()

pclass  survived                             name     sex      age  sibsp  \
1    1st         1    Allen, Miss. Elisabeth Walton  female  29.0000      0   
2    1st         1   Allison, Master. Hudson Trevor    male   0.9167      1   
3    1st         0     Allison, Miss. Helen Loraine  female   2.0000      1   
4    1st         0  Allison, Mr. Hudson Joshua Crei    male  30.0000      1   
5    1st         0  Allison, Mrs. Hudson J C (Bessi  female  25.0000      1   

   parch  ticket        fare    cabin     embarked boat   body  \
1      0   24160  211.337494       B5  Southampton    2    NaN   
2      2  113781  151.550003  C22 C26  Southampton   11    NaN   
3      2  113781  151.550003  C22 C26  Southampton  NaN    NaN   
4      2  113781  151.550003  C22 C26  Southampton  NaN  135.0   
5      2  113781  151.550003  C22 C26  Southampton  NaN    NaN   

                         home.dest  
1                     St Louis, MO  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON  
5  Montreal, PQ / Chesterville, ON

In [7]:
titanic_vectorizer = DictVectorizer(sparse=False)
titanic_feature_cols = ['pclass', 'sex', 'age', 'sibsp']
target_col = 'survived'
titanic_data = titanic_df[[*titanic_feature_cols, target_col]].dropna()
titanic_features = titanic_data[titanic_feature_cols]
X = titanic_vectorizer.fit_transform(titanic_features.to_dict('records'))
titanic_vectorizer.get_feature_names()
y = titanic_data[target_col].values

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=random_state)

In [20]:
clf = MLPClassifier(solver='adam', activation='logistic', alpha=1e-05,
                     hidden_layer_sizes=(2,5), random_state=random_state)
clf.fit(X_train, y_train)

/home/joshua/anaconda3/envs/cpe695/lib/python3.7/site-packages/sklearn/neural_network/_multilayer_perceptron.py:585: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


MLPClassifier(activation='logistic', alpha=1e-05, hidden_layer_sizes=(2, 5),
              random_state=42)

In [21]:
train_survive_indexes = np.where(y_train == 1)[0]
train_survive_predictions = clf.predict(X_train[train_survive_indexes])
inv_train_survive_predictions = np.logical_not(train_survive_predictions).astype(int)
print(f'in-sample survive predicted = {(len(train_survive_predictions) - np.count_nonzero(inv_train_survive_predictions)) / len(train_survive_predictions) * 100:.3f}%')

train_fatality_indexes = np.where(y_train == 0)[0]
train_fatality_predictions = clf.predict(X_train[train_fatality_indexes])
print(f'in-sample fatalities predicted = {(len(train_fatality_predictions) - np.count_nonzero(train_fatality_predictions)) / len(train_fatality_predictions) * 100:.3f}%')

test_survive_indexes = np.where(y_test == 1)[0]
test_survive_predictions = clf.predict(X_test[test_survive_indexes])
inv_test_survive_predictions = np.logical_not(test_survive_predictions).astype(int)
print(f'out-of-sample survive predicted = {(len(test_survive_predictions) - np.count_nonzero(inv_test_survive_predictions)) / len(test_survive_predictions) * 100:.3f}%')

test_fatality_indexes = np.where(y_test == 0)[0]
test_fatality_predictions = clf.predict(X_test[test_fatality_indexes])
print(f'out-of-sample fatalities predicted = {(len(test_fatality_predictions) - np.count_nonzero(test_fatality_predictions)) / len(test_fatality_predictions) * 100:.3f}%')

in-sample survive predicted = 55.193%
in-sample fatalities predicted = 96.192%
out-of-sample survive predicted = 51.111%
out-of-sample fatalities predicted = 96.667%


In [1]:
print('different hidden layer sizes')

different hidden layer sizes


In [14]:
clf = MLPClassifier(solver='adam', activation='logistic', alpha=1e-05,
                     hidden_layer_sizes=(5,4), random_state=random_state)
clf.fit(X_train, y_train)

/home/joshua/anaconda3/envs/cpe695/lib/python3.7/site-packages/sklearn/neural_network/_multilayer_perceptron.py:585: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


MLPClassifier(activation='logistic', alpha=1e-05, hidden_layer_sizes=(5, 4),
              random_state=42)

In [21]:
train_survive_indexes = np.where(y_train == 1)[0]
train_survive_predictions = clf.predict(X_train[train_survive_indexes])
inv_train_survive_predictions = np.logical_not(train_survive_predictions).astype(int)
print(f'in-sample survive predicted = {(len(train_survive_predictions) - np.count_nonzero(inv_train_survive_predictions)) / len(train_survive_predictions) * 100:.3f}%')

train_fatality_indexes = np.where(y_train == 0)[0]
train_fatality_predictions = clf.predict(X_train[train_fatality_indexes])
print(f'in-sample fatalities predicted = {(len(train_fatality_predictions) - np.count_nonzero(train_fatality_predictions)) / len(train_fatality_predictions) * 100:.3f}%')

test_survive_indexes = np.where(y_test == 1)[0]
test_survive_predictions = clf.predict(X_test[test_survive_indexes])
inv_test_survive_predictions = np.logical_not(test_survive_predictions).astype(int)
print(f'out-of-sample survive predicted = {(len(test_survive_predictions) - np.count_nonzero(inv_test_survive_predictions)) / len(test_survive_predictions) * 100:.3f}%')

test_fatality_indexes = np.where(y_test == 0)[0]
test_fatality_predictions = clf.predict(X_test[test_fatality_indexes])
print(f'out-of-sample fatalities predicted = {(len(test_fatality_predictions) - np.count_nonzero(test_fatality_predictions)) / len(test_fatality_predictions) * 100:.3f}%')

in-sample survive predicted = 66.172%
in-sample fatalities predicted = 90.180%
out-of-sample survive predicted = 63.333%
out-of-sample fatalities predicted = 85.000%


### Results

|                      | in sample survive | in sample fatalities | out of sample survive | out of sample fatalities |
|----------------------|-------------------|----------------------|-----------------------|--------------------------|
| pruned decision tree | 56.380%           | 97.595%              | 48.889%               | 96.667%                  |
| first ANN            | 55.193%           | 96.192%              | 51.111%               | 96.667%                  |
| second ANN           | 66.172%           | 90.180%              | 63.333%               | 85.000%                  |

The trained ANN's have comporable performance to the pruned decision tree. By changing the size of the hidden layers you can achieve better performance for fatalities vs survive classifications.